### PR TITLE
sync: another refactoring pass

### PR DESCRIPTION
**Pull request by ee7**
Not everything here is strictly better, but it's hopefully in the right direction overall.

I've got lots more tidying up in the pipeline - the uglier changes shouldn't be here in the long-term.

This PR is part of a series that addresses #64.

### CODE REVIEW

File: src/sync/exercises.nim [hunk 1]
```diff
@@ -1,8 +1,8 @@
-import std/[algorithm, json, options, os, sequtils, sets, strformat, strutils,
-            tables]
+import std/[json, options, os, sequtils, sets, strformat, strutils, tables]
 import pkg/parsetoml
 import ".."/cli
 import "."/[probspecs, tracks]
+export tracks.`$`
```
Comment (ee7) [inline src/sync/exercises.nim:5]: This will probably be refactored away later.

File: tests/test_tracks.nim [hunk 2]
```diff
@@ -52,7 +52,7 @@ proc main =
         )
 
       check:
-        trackExercises[1] == expectedTwoFer
+        trackExercises[65] == expectedTwoFer
```
Comment (ee7) [inline tests/test_tracks.nim:55]: I don't love this, but I guess it's fine.

File: src/sync/exercises.nim [hunk 3]
```diff
@@ -68,7 +69,7 @@ proc findExercises*(conf: Conf): seq[Exercise] =
   let probSpecsExercises = findProbSpecsExercises(conf).mapIt((it.slug, it)).toTable
 
   for practiceExercise in findPracticeExercises(conf):
-    result.add(initExercise(practiceExercise, probSpecsExercises.getOrDefault(practiceExercise.slug)))
+    result.add(initExercise(practiceExercise, probSpecsExercises.getOrDefault(practiceExercise.slug.string)))
```
Comment (ee7) [inline src/sync/exercises.nim]: Most of the things that use `.string` should be refactored away entirely, or become a `{.borrow.}`

File: src/sync/exercises.nim (hunk 1 above)
Comment (ErikSchierboom) [reply src/sync/exercises.nim:5]: Good. It looks slightly odd :)

File: src/sync/exercises.nim [hunk 4]
```diff
@@ -47,29 +48,48 @@ proc newExerciseTestCase(testCase: ProbSpecsTestCase): ExerciseTestCase =
...
+
+proc initExerciseTestCases(testCases: seq[ProbSpecsTestCase]): seq[ExerciseTestCase] =
+  result = newSeq[ExerciseTestCase](testCases.len)
+  for i, testCase in testCases:
+    result[i] = newExerciseTestCase(testCase)
 
-  let reimplementations = testCases.filterIt(it.isReimplementation).mapIt((it.uuid, it.reimplements)).toTable()
-  let testCasesByUuids = result.newTableFrom(proc (testCase: ExerciseTestCase): string = testCase.uuid)
+  let reimplementations = getReimplementations(testCases)
+  let testCasesByUuids =  uuidToTestCase(result)
```
Comment (ErikSchierboom) [inline src/sync/exercises.nim:66]: Was the space intentional?

```suggestion
  let testCasesByUuids = uuidToTestCase(result)
```

**Review by ErikSchierboom — APPROVED** [review]

File: src/sync/exercises.nim (hunk 4 above)
Comment (ee7) [reply src/sync/exercises.nim]: Wasn't intentional - thanks.

**Pull request merged**